Support for readable streams in object mode.

Letrab opened the pull request:
I was using your library with a stream in object mode (for which it is important to keep a blob of data (object) together).

In this case a `stream.read(size)` ignores its `size` parameter, and just returns one object.
(Quoted documentation: _A Readable stream in object mode will always return a single item from a call to readable.read(size), regardless of the value of the size argument._)

In this case, it is likely that the partition size is larger than exactly specified. Do end the current write stream in this case as well.

Bartel-C8 committed:
Support for readable streams in object mode.

Add err to callback

dannycho7 commented:
Thanks for pointing this out. It would be nice to include a test case as well.

I've added comments.

dannycho7 commented on index.js:
```diff
@@ -57,7 +58,8 @@ const _splitToStream = (outStreamCreate, fileStream, partitionStreamSize, callba
 
-			if(currentFileSize == partitionStreamSize) {
+			// In objectMode one object is read from the stream, it could be that the size is (a bit) bigger than the partition size
+			if(currentFileSize >= partitionStreamSize) {
```
I think we should make sure to continue to enforce an actual "maxFileSize". To do this, we can allow files to be slightly smaller than the max file size by adding the object to the next output stream if it could exceed the current output stream.

One edge case to consider is what happens if the maxFileSize is chosen such that 1 object cannot fit inside an output file (in that case, we should throw noting that the user input was invalid).

Bartel-C8 committed:
(deep)equal is deprecated, use strictEqual instead.

Letrab commented:
Did elaborate and add tests (there was also the case where objectMode was used as input for the readable, but the streams was being Transformed)

In order to be able to throw when an object did not fit in maxFileSize, I had to pass the error to the callback function. As when throwing in the .on("readable") I am unable to catch it due to scoping. 
I've added now the error parameter as 2nd in the callback, to not break API. But actually it is (node-style) preferred to pass it as fist param: https://thenodeway.io/posts/understanding-error-first-callbacks/

What do you think?
Will change documentation accordingly after agreeing upon.

Letrab commented:
And actually, it would be nice to add an option (for objectMode) to be able to increase maxFileSize without throwing errors? (As e.g. it could be possible the object size is not known in advance)

dannycho7 commented:
> Did elaborate and add tests (there was also the case where objectMode was used as input for the readable, but the streams was being Transformed)
> 
> In order to be able to throw when an object did not fit in maxFileSize, I had to pass the error to the callback function. As when throwing in the .on("readable") I am unable to catch it due to scoping.
> I've added now the error parameter as 2nd in the callback, to not break API. But actually it is (node-style) preferred to pass it as fist param: https://thenodeway.io/posts/understanding-error-first-callbacks/
> 
> What do you think?
> Will change documentation accordingly after agreeing upon.

This is awesome! Yeah, I don't mind if you change the API. I didn't have the foresight a few years ago. The release would just be a more obvious increment.

dannycho7 commented:
> And actually, it would be nice to add an option (for objectMode) to be able to increase maxFileSize without throwing errors? (As e.g. it could be possible the object size is not known in advance)

I'd question what they're using this API for haha. What'd be the expected behavior if the objects are of size 1 GB each but the maxFileSize is set to 128 bytes?

Bartel-C8 committed:
Change split API to be error-first.

Update version to 2.0.0 as API changed.
Add extra keywords.

Letrab commented:
> > Did elaborate and add tests (there was also the case where objectMode was used as input for the readable, but the streams was being Transformed)
> > In order to be able to throw when an object did not fit in maxFileSize, I had to pass the error to the callback function. As when throwing in the .on("readable") I am unable to catch it due to scoping.
> > I've added now the error parameter as 2nd in the callback, to not break API. But actually it is (node-style) preferred to pass it as fist param: https://thenodeway.io/posts/understanding-error-first-callbacks/
> > What do you think?
> > Will change documentation accordingly after agreeing upon.
> 
> This is awesome! Yeah, I don't mind if you change the API. I didn't have the foresight a few years ago. The release would just be a more obvious increment.

Done! Took the liberty of increasing the version to 1.4.0. Added some keywords as well (took a while for me to find this library)

But maybe a 2.0.0 version is more appropriate, as API changes are breaking (following SemVer: https://zellwk.com/blog/semantic-versioning/)?

Letrab commented:
> > And actually, it would be nice to add an option (for objectMode) to be able to increase maxFileSize without throwing errors? (As e.g. it could be possible the object size is not known in advance)
> 
> I'd question what they're using this API for haha. What'd be the expected behavior if the objects are of size 1 GB each but the maxFileSize is set to 128 bytes?

I was thinking about the reverse case:
Objects of 101kb for a maxFileSize of 100kB e.g. But probably an unlikely case indeed ;-).

dannycho7 commented:
> > > Did elaborate and add tests (there was also the case where objectMode was used as input for the readable, but the streams was being Transformed)
> > > In order to be able to throw when an object did not fit in maxFileSize, I had to pass the error to the callback function. As when throwing in the .on("readable") I am unable to catch it due to scoping.
> > > I've added now the error parameter as 2nd in the callback, to not break API. But actually it is (node-style) preferred to pass it as fist param: https://thenodeway.io/posts/understanding-error-first-callbacks/
> > > What do you think?
> > > Will change documentation accordingly after agreeing upon.
> > 
> > 
> > This is awesome! Yeah, I don't mind if you change the API. I didn't have the foresight a few years ago. The release would just be a more obvious increment.
> 
> Done! Took the liberty of increasing the version to 1.4.0. Added some keywords as well (took a while for me to find this library)
> 
> But maybe a 2.0.0 version is more appropriate, as API changes are breaking (following SemVer: https://zellwk.com/blog/semantic-versioning/)?

Yes, 2.0.0 sounds better! Thanks for everything so far :D

Letrab commented:
Done! Thank you for your work on this lib! ;-)

dannycho7 commented:
@Letrab Could you rebase + squash commits? Seems like commit 1 and 3 should be in the same one.

Letrab commented:
> @Letrab Could you rebase + squash commits? Seems like commit 1 and 3 should be in the same one.

Done!

dannycho7 commented on index.js:
```diff
@@ -36,35 +36,53 @@ const _splitToStream = (outStreamCreate, fileStream, partitionStreamSize, callba
+			const writeChunk = isObjectMode ? JSON.stringify(chunk) : chunk
+
+			if ((currentFileSize + writeChunk.length) <= partitionStreamSize) {
+				currentOutStream.write(isObjectMode ? JSON.stringify(chunk) : chunk);
+				currentFileSize += isObjectMode ? JSON.stringify(chunk).length : chunk.length;
```
This can just be writeChunk and writeChunk.length, but I'll do the cleanup myself.

dannycho7 commented:
I appreciate the help! I've merged and will publish.
